src/public_addresses.rs

```diff
@@ -0,0 +1,145 @@
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use std::{collections::HashSet, sync::Arc};
+
+use multiaddr::{Multiaddr, Protocol};
+use parking_lot::RwLock;
+
+use crate::PeerId;
+
+/// Set of the public addresses of the local node.
+///
+/// The format of the addresses stored in the set contain the local peer ID.
+/// This requirement is enforced by the [`PublicAddresses::add_public_address`] method,
+/// that will add the local peer ID to the address if it is missing.
+///
+/// # Note
+///
+/// - The addresses are reported to the identify protocol and are used by other nodes
+///  to establish a connection with the local node.
+///
+/// - Users must ensure that the addresses are reachable from the network.
+#[derive(Debug, Clone)]
+pub struct PublicAddresses {
+    pub(crate) inner: Arc<RwLock<HashSet<Multiaddr>>>,
+    local_peer_id: PeerId,
+}
+
+impl PublicAddresses {
+    /// Creates new [`PublicAddresses`] from the given peer ID.
+    pub(crate) fn new(local_peer_id: PeerId) -> Self {
+        Self {
+            inner: Arc::new(RwLock::new(HashSet::new())),
+            local_peer_id,
+        }
+    }
+
+    /// Add a public address to the list of addresses.
+    ///
+    /// The address must contain the local peer ID, otherwise an error is returned.
+    /// In case the address does not contain any peer ID, it will be added.
+    ///
+    /// Returns true if the address was added, false if it was already present.
+    pub fn add_public_address(&self, address: Multiaddr) -> Result<bool, Multiaddr> {
+        let address = self.check_local_peer(address)?;
+        Ok(self.inner.write().insert(address))
+    }
+
+    /// Remove the exact public address.
+    ///
+    /// The provided address must contain the local peer ID.
+    pub fn remove_public_address(&self, address: &Multiaddr) -> bool {
+        self.inner.write().remove(address)
+    }
+
+    /// Returns a vector of the available listen addresses.
+    pub fn get_addresses(&self) -> Vec<Multiaddr> {
+        self.inner.read().iter().cloned().collect()
+    }
+
+    /// Checks if the address contains the local peer ID.
+    fn check_local_peer(&self, mut address: Multiaddr) -> Result<Multiaddr, Multiaddr> {
```

Comment:
```suggestion
    fn ensure_local_peer(&self, mut address: Multiaddr) -> Result<Multiaddr, Multiaddr> {
```
As we also append the local peer ID?